Mathlib/Topology/CWComplex.lean

```diff
@@ -0,0 +1,123 @@
+/-
+Copyright (c) 2024 Elliot Dean Young and Jiazhen Xia. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiazhen Xia, Elliot Dean Young
+-/
+
+import Mathlib.Topology.Category.TopCat.Limits.Basic
+import Mathlib.CategoryTheory.Functor.OfSequence
+
+/-!
+# CW-complexes
+
+This file defines (relative) CW-complexes.
+
+## Main definitions
+
+* `RelativeCWComplex`: A relative CW-complex is the colimit of an expanding sequence of subspaces
+`sk i` (called the `(i-1)`-skeleton) for `i ≥ 0`, where `sk 0` (i.e., the `(-1)`-skeleton) is an
+arbitrary topological space, and each `sk (n+1)` (i.e., the `n`-skeleton) is obtained from `sk n`
+(i.e., the `(n-1)`-skeleton) by attaching `n`-disks.
+
+* `CWComplex`: A CW-complex is a relative CW-complex whose `sk 0` (i.e., `(-1)`-skeleton) is empty.
+
+## References
+
+The definition of CW-complexes follows David Wärn's suggestion at
+https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Do.20we.20have.20CW.20complexes.3F/near/231769080
+-/
+
+open CategoryTheory TopCat
+
+universe u
+
+namespace RelativeCWComplex
+
+/-- The inclusion map from the `n`-sphere to the `(n+1)`-disk -/
+def sphereInclusion (n : ℤ) : 𝕊 n ⟶ 𝔻 (n + 1) where
+  toFun := fun ⟨p, hp⟩ ↦ ⟨p, le_of_eq hp⟩
+  continuous_toFun := ⟨fun t ⟨s, ⟨r, hro, hrs⟩, hst⟩ ↦ by
+    rw [isOpen_induced_iff, ← hst, ← hrs]
+    tauto⟩
+
+variable {S D : TopCat.{u}} (f : S ⟶ D)
+
+/-- Given the inclusion map `f : S ⟶ D` for one generalized cell, we construct the inclusion map
+from the disjoint union of `S` (boundary of generalized cells) to the disjoint union of `D`
+(generalized cells), where both of the disjoint unions are indexed by `cells`. -/
+def generalizedSigmaSphereInclusion (cells : Type) :
```

Comment:
I believe that all occurrences of `Type` should be replaced by `Type u`. Otherwise, for example, a discrete topological space in `Type u` may not be a CW-complex.

Reply:
Also, as you are using the colimit API for pushouts below, it would be more consistent to use the general category theory coproducts API instead of the type theoretic disjoint union. This would also remove the need for the verification of the continuity of some maps.

Reply:
Thank you for these comments! Yes, using the categorical coproduct API has simplified the code, and is more consistent.